client/widgets/CArtifactsOfHeroBackpack.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * CArtifactsOfHeroBackpack.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+#include "CArtifactsOfHeroBackpack.h"
+
+#include "../gui/CGuiHandler.h"
+#include "../gui/Shortcut.h"
+
+#include "GameConstants.h"
+#include "Buttons.h"
+#include "GameSettings.h"
+#include "IHandlerBase.h"
+
+#include "../CPlayerInterface.h"
+
+#include "../../CCallback.h"
+
+CArtifactsOfHeroBackpack::CArtifactsOfHeroBackpack(const Point & position, DestroyHandler destroyThisCallback)
+{
+	OBJECT_CONSTRUCTION_CAPTURING(255 - DISPOSE);
+	pos += position;
+
+	const auto backpackCap = VLC->settings()->getInteger(EGameSettings::HEROES_BACKPACK_CAP);
+	auto visibleCapasityMax = GameConstants::HERO_BACKPACK_WINDOW_SLOT_LINES * GameConstants::HERO_BACKPACK_WINDOW_SLOT_COLUMNS;
+	if(backpackCap >= 0)
+		visibleCapasityMax = visibleCapasityMax > backpackCap ? backpackCap : visibleCapasityMax;
+
+	backpack.resize(visibleCapasityMax);
+	size_t artPlaceIdx = 0;
+	for(auto & artPlace : backpack)
+	{
+		artPlace = std::make_shared<CHeroArtPlace>(
+			Point(46 * (artPlaceIdx % GameConstants::HERO_BACKPACK_WINDOW_SLOT_COLUMNS), 46 * (artPlaceIdx / GameConstants::HERO_BACKPACK_WINDOW_SLOT_COLUMNS)));
+		artPlace->setArtifact(nullptr);
+		artPlace->leftClickCallback = std::bind(&CArtifactsOfHeroBase::leftClickArtPlace, this, _1);
+		artPlace->rightClickCallback = std::bind(&CArtifactsOfHeroBase::rightClickArtPlace, this, _1);
+		artPlaceIdx++;
+	}
+	if(backpackCap < 0 || visibleCapasityMax < backpackCap)
+	{
+		auto scrollHandler = std::bind(&CArtifactsOfHeroBase::scrollBackpack, this, _1);
+		leftBackpackRoll = std::make_shared<CButton>(Point(-20, 0), "hsbtns3.def", CButton::tooltip(), [scrollHandler]() { scrollHandler(-1); }, EShortcut::MOVE_LEFT);
+		rightBackpackRoll = std::make_shared<CButton>(Point(368, 318), "hsbtns5.def", CButton::tooltip(), [scrollHandler]() { scrollHandler(+1); }, EShortcut::MOVE_RIGHT);
+		leftBackpackRoll->block(true);
+		rightBackpackRoll->block(true);
```

Comment:
I assume scroll buttons on 8x8 backpack window would scroll backpack by 1 item? Not by 1 row?

IMO we should scroll by 1 row and possibly use slider in place of buttons

Reply:
It is possible to scroll to any number of items. Including on HERO_BACKPACK_WINDOW_SLOT_COLUMNS. But I will also make scrolling a slider.